src/goto-programs/goto_coverage.cpp

```diff
@@ -38,64 +62,666 @@ void goto_coveraget::add_false_asserts(goto_functionst &goto_functions)
+
+  *top_ptr = join_rev_expr;
+  if (pre_cond_flg)
+  {
+    exprt not_pre_cond = exprt("not", bool_type());
+    not_pre_cond.operands().emplace_back(root_ptr->op0());
+    root_ptr->op0() = not_pre_cond;
+    root_ptr->id(exprt::id_and);
+  }
+
+  join_not_expr = exprt("not", bool_type());
+  join_not_expr.operands().emplace_back(*root_ptr);
+
+  migrate_expr(join_not_expr, guard);
+  migrate_expr(rhs, a_guard);
+
+  t = goto_program.insert(it);
+  t->type = ASSERT;
+  t->guard = guard;
+  t->location = it->location;
+  t->location.property("instrumented assertion");
+  t->location.comment(from_expr(ns, "", a_guard));
+  t->location.user_provided(true);
+  it = ++t;
+
+  idf = from_expr(ns, "", a_guard) + "\t" + it->location.as_string();
+
+  // 7. restore root_ptr
+  // noted that this should be done before the updating of *top_ptr
+  if (pre_cond_flg)
+  {
+    // change 'not (a && a) and b' back to 'a && a or b'
+    *root_ptr = old_root;
+  }
+
+  // 8. update *top_ptr;
+  join_expr.clear();
+  *top_ptr = old_top;
+  join_expr = exprt(op_tp, bool_type());
+  join_expr.operands().emplace_back(*top_ptr);
+  join_expr.operands().emplace_back(rhs);
+  *top_ptr = join_expr;
+
+  // 9. update rhs_ptr
+  rhs_ptr = top_ptr->operands().begin();
+  rhs_ptr++;
+}
+
+/*
+  - flag: if we have handled one
+*/
+void goto_coveraget::collect_operands(
+  const exprt &expr,
+  std::list<exprt> &operands,
+  bool &flag)
+{
+  const std::string &id = expr.id().as_string();
+
+  if ((id == "and" || id == "or") && flag == false)
+  {
+    bool flg0 = false;
+    collect_operands(expr.op0(), operands, flg0);
+    if (!flg0)
+      operands.emplace_back(expr.op0());
+    // operators.emplace_back(id);
+    bool flg1 = false;
+    collect_operands(expr.op1(), operands, flg1);
+    if (!flg1)
+      operands.emplace_back(expr.op1());
+    flag = true;
+  }
+  else
+  {
+    forall_operands (it, expr)
+      collect_operands(*it, operands, flag);
+  }
+}
+
+void goto_coveraget::collect_atom_operands(
+  const exprt &expr,
+  std::set<exprt> &atoms)
+{
+  const auto &id = expr.id();
+  forall_operands (it, expr)
+    collect_atom_operands(*it, atoms);
+  if (
+    id == exprt::equality || id == exprt::notequal || id == exprt::i_lt ||
+    id == exprt::i_gt || id == exprt::i_le || id == exprt::i_ge)
+  {
+    atoms.insert(expr);
+  }
+}
+
+void goto_coveraget::collect_operators(
+  const exprt &expr,
+  std::list<std::string> &operators)
+{
+  std::string str = from_expr(expr);
+  // prepocess: remove parenthesis in type_cast
+  // e.g. if(return_bool()) ==> IF !(_Bool)return_value$_return_bool$1
+  //                        ==> IF !return_value$_return_bool$1
+  // !FIXME: maybe this should be regex like \(\_[A-Za-z]+\) iF there are other typecasts
+  const std::string subStr = "(_Bool)";
+  size_t pos = str.find(subStr);
+  while (pos != std::string::npos)
+  {
+    str.erase(pos, subStr.length());
+    pos = str.find(subStr, pos);
+  }
```

Comment:
Hard-coded C assumption. Typically, in ESBMC, if we want to not process (certain) typecasts, there would be a recursive function applied to an expression tree, which replaces those typecasts with the result of the recursive function applied to the child node. Depending on what the code below does (I see lots of string manipulation), it might be a good idea to implement such a recursive function modifying an expression tree.

Reply:
I agree. But it might be time-consuming to do that. I am afraid I cannot implement that in this PR.